src/state/messages/message-drafts.tsx

```diff
@@ -0,0 +1,84 @@
+import React, {useEffect, useMemo, useReducer, useRef} from 'react'
+
+import {useCurrentConvoId} from './current-convo-id'
+
+const MessageDraftsContext = React.createContext<{
+  state: State
+  dispatch: React.Dispatch<Actions>
+} | null>(null)
+
+function useMessageDraftsContext() {
+  const ctx = React.useContext(MessageDraftsContext)
+  if (!ctx) {
+    throw new Error(
+      'useMessageDrafts must be used within a MessageDraftsContext',
+    )
+  }
+  return ctx
+}
+
+export function useMessageDraft() {
+  const {currentConvoId} = useCurrentConvoId()
+  const {state, dispatch} = useMessageDraftsContext()
+  return useMemo(
+    () => ({
+      getDraft: () => (currentConvoId && state[currentConvoId]) || '',
+      clearDraft: () => {
+        if (!currentConvoId) {
+          throw new Error('No current convo id')
+        }
+        dispatch({type: 'clear', convoId: currentConvoId})
+      },
+    }),
+    [state, dispatch, currentConvoId],
+  )
+}
+
+export function useSaveMessageDraft(message: string) {
+  const {currentConvoId} = useCurrentConvoId()
+  const {dispatch} = useMessageDraftsContext()
+  const messageRef = useRef(message)
+  messageRef.current = message
+
+  useEffect(() => {
+    return () => {
+      if (currentConvoId) {
+        dispatch({
+          type: 'set',
+          convoId: currentConvoId,
+          draft: messageRef.current,
+        })
+      }
+    }
+  }, [currentConvoId, dispatch])
+}
```

Comment:
This is the sketchy part. Can't think of another way to stop `useEffect` firing every time the message changes (other than a completely different approach like event listeners)